firedrake/adjoint/ensemble_reduced_functional.py

```diff
@@ -53,15 +61,72 @@ class EnsembleReducedFunctional(ReducedFunctional):
     works, please refer to the `Firedrake manual
     <https://www.firedrakeproject.org/parallelism.html#id8>`_.
     """
-    def __init__(self, J, control, ensemble, scatter_control=True):
-        super(EnsembleReducedFunctional, self).__init__(J, control)
+    def __init__(self, J, control, ensemble, scatter_control=True,
+                 gather_functional=None):
         self.ensemble = ensemble
+        if isinstance(J, list):
+            assert isinstance(control, list), "Controls should also be a list."
+            assert len(control) % len(J) == 0, "Controls and J have mismatching lengths."
+            self.controls_per_J = len(control) // len(J)
+            self.functional = J
+            self.controls = Enlist(control)
+            self.Jhats = []
+            self.sizes = self.ensemble.ensemble_comm.allgather(len(J))
+            for i in range(len(J)):
+                i0 = self.controls_per_J*i
+                i1 = self.controls_per_J*(i+1)
+                self.Jhats.append(ReducedFunctional(J[i], control[i0:i1]))
+        else:
+            super(EnsembleReducedFunctional, self).__init__(J, control)
         self.scatter_control = scatter_control
+        self.gather_functional = gather_functional
+
+    def _allgather_J(self, J):
+        if isinstance(J, list):
+            rank = self.ensemble.ensemble_comm.rank
+            vals = []
+            # allgather a flattened list of all of the functional values
+            for i in range(len(self.sizes)):
+                for j in range(self.sizes[i]):
```

Comment:
It is better to use `enumerate` here, e.g.:

```
for i, size in enumerate(self.sizes):
    for j in range(size):
```

Reply:
loop over the Js directly

Reply:
I can't here, because each J is local to one ensemble rank.